libraries/standard/mqtt/src/mqtt.c

```diff
@@ -163,7 +163,32 @@ MQTTStatus_t MQTT_Unsubscribe( MQTTContext_t * const pContext,
 
 MQTTStatus_t MQTT_Disconnect( MQTTContext_t * const pContext )
 {
-    return MQTTSuccess;
+    size_t packetSize;
+    int32_t bytesSent;
+
+    MQTTStatus_t status = MQTT_GetDisconnectPacketSize( &packetSize );
+
+    if( status == MQTTSuccess )
+    {
+        status = MQTT_SerializeDisconnect( &( pContext->networkBuffer ) );
+    }
+
+    if( status == MQTTSuccess )
+    {
+        bytesSent = sendPacket( pContext, packetSize );
+
+        if( bytesSent < 0 )
+        {
+            status = MQTTSendFailed;
```

Comment:
Per the spec,

After sending a DISCONNECT Packet the Client:
* MUST close the Network Connection[MQTT-3.14.4-1].
* MUST NOT send any more Control Packets on that Network Connection[MQTT-3.14.4-2].

Should we even be checking the whether we were able to send the disconnect packet successfully?

Reply:
I think so.  The application should know if the disconnect failed to send and can decide whether it is necessary to try again.  Closing the network connection without providing assurance of the Disconnect being sent may cause some unintended side effects (for example, an unexpected Last Will and Testament message being published).

Reply:
Would `bytesSent` contain a lower-level error code when it's `< 0`?  If so, should we pass it upwards so that it can be decoded and logged?  How will someone debug a failure when this happens?

Reply:
This is an open discussion beyond the scope of this PR.  See comment from @abhidixi11 below:

> I'm wondering about MQTTStatus_t here. Should we have one Network Error status rather than send fail or receive fail. The more specific error code can be communicated via context, (error code bubbling up, I'm still working on that proposal as part of transport layer ).